08-optional-types.Rmd

```diff
@@ -0,0 +1,344 @@
+
+```
+# this workflow will fail miniwdl check or womtool
+workflow do_x_or_y {
+    input {
+        Boolean do_x = true # if true do task_x, if false do task_y
+        File some_input
+    }
+    if (do_x = true) {
+        Int some_variable = 10  # this is problematic, even though only one of these if blocks will execute!
+        call task_x {
+            input:
+                some_input = some_input
+        }
+    }
+    if (do_x = false) {
+        Int some_variable = 5  # this is problematic, even though only one of these if blocks will execute!
+        call task_y {
+         input:
+                some_input = some_input
+        }
+    }
+}
+```
+Another consequence is that both of your mutually exclusive tasks' outputs will be considered optional types, even though you can be certain that set of outputs absolutely do exist and one set of outputs absolutely do not exist.
+
+```
+workflow do_x_or_y {
+    input {
+        Boolean do_x = true # if true do task_x, if false do task_y
+        File some_input
+    }
+    if (do_x = true) {
+        call task_x {
+            input:
+                some_input = some_input
+        }
+    }
+    if (do_x = false) {
+        call task_y {
+         input:
+                some_input = some_input
+        }
+    }
+
+    Boolean did_x_run = defined(task_x.something)
+    Boolean did_y_run = defined(task_y.something)
+
+    outputs {
+        # these both MUST be optional
+        File? task_x_output = task_x.something
+        File? task_y_output = task_y.something
+    }
+}
+```
+
+## Checking if a variable is defined
+You can check if a variable is defined at runtime using the WDL built-in `defined()`, which returns a Boolean value. However, `defined()` can give unexpected output, as we will go over later. For the time being, let's use a simple example.
+
+```
+workflow do_x_or_y {
+    input {
+        Boolean do_x = true # if true do task_x, if false do task_y
+        File some_input
+    }
+    if (do_x = true) {
+        call task_x {
+            input:
+                some_input = some_input
+        }
+    }
+    if (do_x = false) {
+        call task_y {
+         input:
+                some_input = some_input
+        }
+    }
+
+    # these work as you would expect
+    Boolean did_x_run = defined(task_x.something)
+    Boolean did_y_run = defined(task_y.something)
+
+    outputs {
+        File? task_x_output = task_x.something
+        File? task_y_output = task_y.something
+        Boolean did_x_run = did_x_run
+        Boolean did_y_run = did_y_run
+    }
+}
+```
+
+Be aware that WDL is not "aware" of any of these checks. Even if your workflow has established a variable is defined, WDL does not know that, so you cannot use these methods to use an optional variable like a non-optional variable.
```

Comment:
Your second sentence in this paragraph lost me...

Reply:
This doesn't work:
```
workflow yargh {
  input { 
    Int? optional_variable 
  }
  if(defined(optional_variable) {
    # if we are in this if-block, optional_variable must be defined
    # nevertheless, this will cause an error because WDL doesn't "know" that
    call some_task { input: required_variable = optional_variable }
  }
}
```

I think it's best to use an example rather than try to explain it in text... it's tricky!

Reply:
Decided to rework the section, let me know if it's a better option.